[FEATURE] Microsoft FabricPowerBIDatasource

Kilo59 opened the pull request:
Adding experimental `FabricPowerBIDatasource`

## Note
Do not include in release notes

## TODO
- [x] Add `measure` param
- ~Custom serde for `filters` param (tuple for dict keys)~
- [x] check for running inside fabric notebook


### Followup Questions
- Any special Datadocs inside a fabric notebook handling?

Kilo59 committed:
`add_fabric()` stub

Kilo59 committed:
`get_batch_list_from_batch_request()` partial impl

copied from pandas datasource

Kilo59 commented on great_expectations/datasource/fluent/fabric.py:
```diff
@@ -0,0 +1,287 @@
+]
+
+
+class FabricDatasource(Datasource):
```
```suggestion
class FabricPowerBIDatasource(Datasource):
```

Kilo59 commented on great_expectations/datasource/fluent/fabric.py:
```diff
@@ -0,0 +1,287 @@
+
+    # right side of the operator determines the type name
+    # left side enforces the names on instance creation
+    type: Literal["fabric"] = "fabric"
```
```suggestion
    type: Literal["fabric_powerbi"] = "fabric_powerbi"
```

Kilo59 commented on great_expectations/datasource/fluent/fabric.py:
```diff
@@ -0,0 +1,287 @@
+    ]
+    # any fabric datsource specific fields should be added to this set
+    # example a connection_string field or a data directory field
+    _EXTRA_EXCLUDED_EXEC_ENG_ARGS: ClassVar[set] = {"extra_field_1"}
```
```suggestion
    _EXTRA_EXCLUDED_EXEC_ENG_ARGS: ClassVar[set] = {"workspace", "dataset"}
```

ghost commented:
<details open>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=8605&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations">Click on the image</a> for a new way to code review</b>
</summary>

####

[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/great-expectations/great_expectations/8605/526cb93c/1b8828adfcbb5bfec944e8d8f5d3cd190a0c40a3.svg)](https://app.codesee.io/r/reviews?pr=8605&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)

#### Legend
<img src="https://app.codesee.io/Legend_v3.png" alt="CodeSee Map legend" width="220" height="156">
</details>

Kilo59 committed:
make asset `test_connection()` a noop

Kilo59 committed:
rename to `FabricPowerBI`

Kilo59 committed:
add asset fields

Kilo59 committed:
PandasExecutionEngine updates

update `.get_batch_data_and_markers()` types
skip `_apply_splitting_and_sampling` with instance check

Kilo59 committed:
initial scaffolding

Kilo59 committed:
ignore/fix mypy warning

Kilo59 committed:
update schemas

Kilo59 committed:
todos

Kilo59 committed:
use `"fluent"` data_connector_name

Kilo59 committed:
add example yaml config

Kilo59 committed:
update stubs

Kilo59 committed:
TDD test and fixes

Kilo59 committed:
fix reader options

Kilo59 committed:
add remaining parameters

Kilo59 committed:
add missing `measure` param

Kilo59 committed:
Merge branch 'develop' into ms-fabric

Kilo59 commented on great_expectations/execution_engine/pandas_execution_engine.py:
```diff
@@ -367,21 +373,26 @@ def get_batch_data_and_markers(  # noqa: C901, PLR0912, PLR0915
+        batch_spec: BatchSpec | PandasBatchSpecProtocol,
+        batch_data: PandasBatchData,
+    ):
+        if isinstance(batch_spec, BatchSpec):
+            splitter_method_name: Optional[str] = batch_spec.get("splitter_method")
```
Splitting and sampling has not been implemented. Let's skip this for the first iteration.
Also, this isn't the best way to deal with this issue.

eisber replied:
is splitting like paging?

except for read_table, adding sampling can't be done generically.

Kilo59 replied:
I would say it's more like partitioning the data based on some category, like by the directory it's stored in or a date range.

Kilo59 committed:
Merge branch 'develop' into ms-fabric

Kilo59 committed:
cleanup

Kilo59 committed:
ensure running on fabric

Kilo59 committed:
move to experimental directory

Kilo59 committed:
type ignore

Kilo59 commented on great_expectations/experimental/datasource/fabric.py:
```diff
@@ -0,0 +1,396 @@
+                "Could not import `sempy.fabric`\npip install semantic-link"
+            ) from import_err
+
+        # TODO: check if we are running from within fabric?
```
```suggestion
```

Kilo59 committed:
remove comment

Kilo59 committed:
update tests

Kilo59 committed:
update schemas

Kilo59 committed:
fix doc checks

Kilo59 committed:
Merge branch 'develop' into ms-fabric

Kilo59 committed:
Merge remote-tracking branch 'origin/ms-fabric' into ms-fabric

Kilo59 committed:
use compatibility import

eisber commented on great_expectations/datasource/fluent/schemas/FabricPowerBIDatasource.json:
```diff
@@ -0,0 +1,314 @@
+                },
+                "num_rows": {
+                    "title": "Num Rows",
+                    "type": "integer"
```
this defaults to none

Kilo59 replied:
In latest schema

eisber commented on great_expectations/datasource/fluent/schemas/FabricPowerBIDatasource.json:
```diff
@@ -0,0 +1,314 @@
+                    "type": "string"
+                },
+                "fully_qualified_columns": {
+                    "title": "Fully Qualified Columns",
```
I assume you surface this in some UX. do you also have a long description field?

Kilo59 replied:
Eventually, the schemas may be used to build UI web forms but we are at least a few months away from that.

eisber commented on great_expectations/datasource/fluent/schemas/FabricPowerBIDatasource.json:
```diff
@@ -0,0 +1,314 @@
+                    "default": true,
+                    "type": "boolean"
+                }
+            },
```
dataset and workspace parameter are not included here?

Kilo59 replied:
You may have been looking at an old version of the schema. They are there now.

eisber commented on great_expectations/datasource/fluent/schemas/FabricPowerBIDatasource.json:
```diff
@@ -0,0 +1,314 @@
+                    }
+                },
+                "filters": {
+                    "title": "Filters",
```
Optional[Dict[Tuple[str, str], List[str]]]

the type is a bit complex... here's an example:

            {
                ('State', 'Region'):    ["East", "Central"],
                ('State', 'State'):     ["WA", "CA"]
            }

State = table; Region = Column; East and Central are allowed values for State.Region.

eisber replied:
we're looking into a simplified signature here. would Dict[str, str] work?

Kilo59 replied:
That works.
But if tuples work better we can do that, it just won't be serializable in the initial release.

Kilo59 replied:
@eisber Do you have an updated example?

Something like this?
```python
{
"State.Region": "East.Central",
"State.State": "WA.CA",
}
```

Kilo59 replied:
I've updated this PR to use the simplified `Dict[str, str]` signature.
If we want to revisit tuples as keys in a followup we can.

eisber replied:
Have a look at https://learn.microsoft.com/en-us/python/api/semantic-link-sempy/sempy.fabric?view=semantic-link-python#sempy-fabric-evaluate-measure

 {
       "State[Region]":    ["East", "Central"],
       "State[State]":     ["WA", "CA"]
   }

The right hand-side needs to be a list. it's read like:

Valid combinations are: 

"East", "WA" (so Region = East, State = WA)
"Central",  "CA" (So Region = Central, State = CA)

eisber replied:
ok, I'm story I shared the wrong signature: filters: Dict[str, List[str]]

they keys are not tuples anymore, but the values need to remain lists.

Kilo59 replied:
@eisber Thank you. Updated to make the values lists.

eisber commented on great_expectations/datasource/fluent/schemas/FabricPowerBIDatasource.json:
```diff
@@ -0,0 +1,314 @@
+                "groupby_columns": {
+                    "title": "Groupby Columns",
+                    "type": "array",
+                    "items": {
```
defaults to None

Kilo59 replied:
In latest schema

eisber commented on great_expectations/datasource/fluent/schemas/FabricPowerBIDatasource.json:
```diff
@@ -0,0 +1,314 @@
+                "filters": {
+                    "title": "Filters",
+                    "type": "object",
+                    "additionalProperties": {
```
defaults to None

eisber commented on great_expectations/datasource/fluent/schemas/FabricPowerBIDatasource.json:
```diff
@@ -0,0 +1,314 @@
+                },
+                "fully_qualified_columns": {
+                    "title": "Fully Qualified Columns",
+                    "type": "boolean"
```
so this is a tri-state, none/true/false. each has a specific behavior.

defaults to None

Kilo59 replied:
This is in the latest schema

eisber commented on great_expectations/datasource/fluent/schemas/FabricPowerBIDatasource.json:
```diff
@@ -0,0 +1,314 @@
+                },
+                "num_rows": {
+                    "title": "Num Rows",
+                    "type": "integer"
```
defaults to None

Kilo59 replied:
In latest schema

eisber commented on great_expectations/experimental/datasource/fabric.py:
```diff
@@ -0,0 +1,394 @@
+            from sempy import fabric  # noqa: F401 # test if fabric is installed
+        except Exception as import_err:
+            raise TestConnectionError(
+                "Could not import `sempy.fabric`\npip install semantic-link"
```
you can replace with semantic-link with semantic-link-sempy

Kilo59 replied:
Done

eisber commented on tests/datasource/fluent/test_fabric.py:
```diff
@@ -0,0 +1,143 @@
+                    "measure": "my_measure",
+                    "groupby_columns": [("foo", "bar"), ("fizz", "buzz")],
+                    # TODO: require custom serde for keys that are tuples
+                    # "filters": {
```
filters is not supported for now, right?

Kilo59 replied:
They are supported, they just won't be serializable, initially.
So they would work with an in-memory data-context but they will fail to serialize with a `Cloud` of `File` Data context.

```python
import great_expectations as gx

in_memory_ctx = gx.get_context(mode="ephemeral")
file_ctx = gx.get_context(mode="file")
cloud_ctx = gx.get_context(mode="cloud")
```

Apologies, the effort to serialize/deserialize shouldn't be too great, I just have been focused on other issues

eisber replied:
we changed the tuples to regular strings. would that solve the serialization/deserialization issue?

Kilo59 replied:
Yes.

Edit:
This issue has been resolved.

Kilo59 committed:
Merge branch 'develop' into ms-fabric

Kilo59 committed:
Merge branch 'develop' into ms-fabric

Kilo59 committed:
`semantic-link` -> `semantic-link-py`

Kilo59 committed:
update filters signature

Kilo59 committed:
add splitting and sampling comment

cdkini commented on great_expectations/core/batch_spec.py:
```diff
@@ -90,6 +108,46 @@ def reader_options(self) -> dict:
+        }
+
+    def get_reader_function(self) -> Callable[..., pd.DataFrame]:
+        # lazy import of fabric module which cotains the reader functions
```
Nit - "contains" typo

Kilo59 replied:
Thank you!

cdkini commented on great_expectations/core/batch_spec.py:
```diff
@@ -90,6 +108,46 @@ def reader_options(self) -> dict:
+        # lazy import of fabric module which cotains the reader functions
+        from sempy import fabric
+
+        return getattr(fabric, self.reader_method)
```
Is this guaranteed? Should we try/except it or handle errors?

Kilo59 replied:
I want it to error here if the method doesn't exist.
I did update to provide a clearer error message though.

cdkini commented on great_expectations/execution_engine/pandas_execution_engine.py:
```diff
@@ -353,6 +355,10 @@ def get_batch_data_and_markers(  # noqa: C901, PLR0912, PLR0915
+        elif isinstance(batch_spec, FabricBatchSpec):
+            reader_fn = batch_spec.get_reader_function()
+            df = reader_fn(**batch_spec.reader_options)
+
         else:
             raise gx_exceptions.BatchSpecError(
                 f"""batch_spec must be of type RuntimeDataBatchSpec, PandasBatchSpec, PathBatchSpec, S3BatchSpec, or AzureBatchSpec, \
```
Should we update this exception with the new type?

Kilo59 replied:
Done

cdkini commented on great_expectations/experimental/datasource/fabric.py:
```diff
@@ -0,0 +1,394 @@
+    @override
+    def test_connection(self) -> None:
+        """
+        Whatever is needed to test the connection to and/or validatitly of the asset.
```
Nit - "validity" typo

cdkini commented on great_expectations/experimental/datasource/fabric.py:
```diff
@@ -0,0 +1,394 @@
+        Whatever is needed to test the connection to and/or validatitly of the asset.
+        This could be a noop.
+        """
+        pass
```
Do we want a complete noop or should we log something?

Kilo59 replied:
I'll add a debug level log.

cdkini commented on great_expectations/experimental/datasource/fabric.py:
```diff
@@ -0,0 +1,394 @@
+        self._validate_batch_request(batch_request)
+        batch_list: List[Batch] = []
+
+        reader_options = {
+            "workspace": self._datasource.workspace,
+            "dataset": self._datasource.dataset,
+            **self.dict(
+                exclude=self._EXCLUDE_FROM_READER_OPTIONS,
+                exclude_none=True,
+                exclude_unset=True,
+                by_alias=True,
+                config_provider=self._datasource._config_provider,
+            ),
+        }
```
Does this dict have any schema/structure?

Kilo59 replied:
The structure comes from the Asset model fields.
But at this point validation has already been done and we are just passing it along to the `batch_spec`

cdkini commented on great_expectations/experimental/datasource/fabric.py:
```diff
@@ -0,0 +1,394 @@
+    @staticmethod
+    def _running_on_fabric() -> bool:
+        if (
+            os.environ.get("AZURE_SERVICE", None)  # noqa: TID251 # needed for fabric
+            != "Microsoft.ProjectArcadia"
+        ):
```
I thought we were avoiding `os.environ` usage outside of config files - is there somewhere else we could grab this from?

cdkini replied:
Nit - the `None` is unnecessary and I'd love to have "Microsoft.ProjectArcadia" be a constant (perhaps at the file or class level)

Kilo59 replied:
This is being used to test whether we are running from inside a fabric notebook.
We don't actually use the value for anything else. It doesn't really make sense in a config file.

We aren't referencing the strings elsewhere so the only value in using a constant is to encode semantic meaning to the string (which IMO is valuable I just need to think of a good name).

Edit:
Named it `_REQUIRED_FABRIC_SERVICE` although maybe `_AZURE_FABRIC_SERVICE` is better

cdkini commented on great_expectations/experimental/datasource/fabric.py:
```diff
@@ -0,0 +1,394 @@
+        from pyspark.sql import SparkSession  # noqa: TID251 # needed for fabric
+
+        sc = SparkSession.builder.getOrCreate().sparkContext
+        return sc.getConf().get("spark.cluster.type") != "synapse"
```
Same note here with these magic strings - if we put them as constants, we can add some context to them

Kilo59 replied:
I've added a constant for `"Microsoft.ProjectArcadia"`, but I'm hesitant to do it here.

@eisber I copied the line on `394` from a snippet you provided, should this be `== "synapse"` instead of `!= "synapse"`?

eisber replied:
I copied that code from our own code base. there's Synapse (PaaS) and Fabric (SaaS) and within Fabric there are also things called Synapse. code can't keep up with marketing.

Kilo59 committed:
add `FabricBatchSpec` to error message

Kilo59 committed:
typo

Kilo59 committed:
log `test_connection` noop

Kilo59 committed:
`_REQUIRED_FABRIC_SERVICE` constant

Kilo59 committed:
add link to fabric package docs

Kilo59 committed:
fix signatures

Kilo59 committed:
spelling and try-except

Kilo59 committed:
Merge branch 'develop' into ms-fabric

Kilo59 committed:
Merge branch 'develop' into ms-fabric
